src/main/java/de/tum/in/www1/artemis/repository/DynamicSpecificationRepository.java

```diff
@@ -0,0 +1,64 @@
+package de.tum.in.www1.artemis.repository;
+
+import java.util.Collection;
+import java.util.Optional;
+
+import jakarta.persistence.criteria.JoinType;
+import jakarta.validation.constraints.NotNull;
+
+import org.springframework.data.jpa.domain.Specification;
+
+import de.tum.in.www1.artemis.domain.DomainObject_;
+import de.tum.in.www1.artemis.repository.fetchOptions.FetchOptions;
+import de.tum.in.www1.artemis.web.rest.errors.EntityNotFoundException;
+
+/**
+ * A repository that provides methods for fetching entities dynamically using {@link FetchOptions}.
+ *
+ * @param <T> the type of the entity
+ */
+public interface DynamicSpecificationRepository<T> {
+
+    /**
+     * Find an entity by applying the given specification.
+     * Provided by the concrete repository implementation. Do not override.
+     *
+     * @param spec the specification to apply
+     * @return the entity that matches the specification
+     */
+    Optional<T> findOne(Specification<T> spec);
+
+    /**
+     * Find an entity by its id or throw an EntityNotFoundException if it does not exist.
+     *
+     * @param specification the specification to apply
+     * @param id            the id of the entity to find
+     * @param entityName    the name of the entity to find
+     * @return the entity with the given id
+     */
+    default T findOneByIdElseThrow(Specification<T> specification, long id, String entityName) {
+        final Specification<T> hasIdSpec = (root, query, criteriaBuilder) -> criteriaBuilder.equal(root.get(DomainObject_.ID), id);
+        return findOne(specification.and(hasIdSpec)).orElseThrow(() -> new EntityNotFoundException(entityName, id));
```

Comment:
Can we also dd a `Optional<T> findOneById()` method?

Reply:
That already exists as part of the normal `JpaRepository` under `findById()`, so it would be somewhat redundant.